### PR TITLE
Handle passthrough copied files in Eleventy

### PR DESCRIPTION
**Pull request by uncenter**
re: https://github.com/webpro/knip/pull/454#issuecomment-1898688280

### CODE REVIEW

File: packages/knip/src/plugins/eleventy/index.ts [hunk 1]
```diff
@@ -32,19 +33,39 @@ const findEleventyDependencies: GenericPluginCallback = async (configFilePath, o
...
-    config?.entry ?? [
-      join(inputDir, dataDir, '**/*.js'),
-      join(inputDir, `**/*.{${typeof templateFormats === 'string' ? templateFormats : templateFormats.join(',')}}`),
-      join(inputDir, '**/*.11tydata.js'),
-    ]
-  ).map(toEntryPattern);
+  const copiedPackages = [];
+  const copiedEntries = [];
+
+  for (const path of Object.keys(dummyUserConfig.passthroughCopies).map(normalize)) {
```
Comment (uncenter) [inline packages/knip/src/plugins/eleventy/index.ts:47]: Normalized so we don't have to deal with `./`, `/`, etc.

File: packages/knip/src/plugins/eleventy/index.ts [hunk 2]
```diff
@@ -32,19 +33,39 @@ const findEleventyDependencies: GenericPluginCallback = async (configFilePath, o
...
+    const parts = path.split('/');
+    if (path.startsWith('node_modules/') && parts.length > 1) {
+      copiedPackages.push(parts.at(1) as string);
+    } else {
+      if (path.endsWith('/')) {
+        copiedEntries.push(`${path}*.{${DEFAULT_EXTENSIONS.join(',')}}`);
+      } else copiedEntries.push(path);
```
Comment (uncenter) [inline packages/knip/src/plugins/eleventy/index.ts:54]: If it isn't in node_modules but it is a directory, add a glob that looks for all files with one of the DEFAULT_EXTENSIONS within that directory. If it isn't in node_modules or a directory, add it to our list of entry files.

File: packages/knip/src/plugins/eleventy/index.ts [hunk 3]
```diff
@@ -32,19 +33,39 @@ const findEleventyDependencies: GenericPluginCallback = async (configFilePath, o
...
+    }
+  }
+
+  return [
+    ...(
+      config?.entry ?? [
+        join(inputDir, dataDir, '**/*.js'),
+        join(inputDir, `**/*.{${typeof templateFormats === 'string' ? templateFormats : templateFormats.join(',')}}`),
+        join(inputDir, '**/*.11tydata.js'),
+        ...copiedEntries.filter(path => DEFAULT_EXTENSIONS.includes(extname(path))),
```
Comment (uncenter) [inline packages/knip/src/plugins/eleventy/index.ts:64]: Remove any entry files from passthrough copy that don't have a supported extension.

**Comment by webpro** [comment]
Thanks for the PR, this is so cool! I'm going to review it closely tomorrow.

File: packages/knip/src/plugins/eleventy/index.ts (hunk 1 above)
Comment (webpro) [reply packages/knip/src/plugins/eleventy/index.ts:47]: I'm going to use this particular comment to reply, but it covers a few of the following lines as well. It's a combination of things why I think we can improve this. little:

- paths including `node_modules` don't need to be translated to a package name here, Knip core has ways to deal with that (e.g. `@org/name` is also a valid package name)
- I've seen directories that end with `/` and some that don't
- The `DEFAULT_EXTENSIONS` already include the dot in e.g. `.js`

When fiddling with some cases I ended up with this:

```ts
const exts = DEFAULT_EXTENSIONS.map(extname => extname.slice(1)).join(',');
const copiedPackages = new Set<string>();
const copiedEntries = new Set<string>();

for (const path of Object.keys(dummyUserConfig.passthroughCopies)) {
  const isDir = !path.includes('*') && isDirectory(join(dirname(configFilePath), path));
  if (isDir) {
    copiedEntries.add(join(path, `**/*.{${exts}}`));
  } else if (isInternal(path)) {
    copiedEntries.add(path);
  } else {
    copiedPackages.add(path);
  }
}
```

Eager to learn what you think of this and if this works properly for your use cases well?

File: packages/knip/src/plugins/eleventy/index.ts (hunk 3 above)
Comment (webpro) [reply packages/knip/src/plugins/eleventy/index.ts:64]: The `copiedEntries` we can return unfiltered for a few reasons:

- Users can configure compilers, and they're not part of `DEFAULT_EXTENSIONS`
- Knip will not add files that don't match valid extensions
- Using `includes()` in theory could match anywhere in the path string
- The paths could be globs like `*.{js,ts}` that won't have a match with a default extension

File: packages/knip/test/plugins/eleventy.test.ts [hunk 4]
```diff
@@ -20,5 +20,7 @@ test('Find dependencies in Eleventy configuration (dynamic)', async () => {
     'entry:_data/**/*.js',
     'entry:content/**/*.{md,njk,html,liquid}',
     'entry:content/**/*.11tydata.js',
+    'entry:js/client/script.js',
+    'prismjs',
```
Comment (webpro) [inline packages/knip/test/plugins/eleventy.test.ts:24]: Victory 🎉

**Review by webpro — COMMENTED** [review]
I like where this is going, and think we're almost there!

**Comment by uncenter** [comment]
> ```js
> const exts = DEFAULT_EXTENSIONS.map(extname => extname.slice(1)).join(',');
> const copiedPackages = new Set<string>();
> const copiedEntries = new Set<string>();
> 
> for (const path of Object.keys(dummyUserConfig.passthroughCopies)) {
>   const isDir = !path.includes('*') && isDirectory(join(dirname(configFilePath), path));
>   if (isDir) {
>     copiedEntries.add(join(path, `**/*.{${exts}}`));
>   } else if (isInternal(path)) {
>     copiedEntries.add(path);
>   } else {
>     copiedPackages.add(path);
>   }
> }
> ```

Makes sense to me!

**Comment by webpro** [comment]
Awesome, can I merge it?

**Review by uncenter — COMMENTED** [review]
LGTM!

**Comment by webpro** [comment]
:rocket: _This pull request is included in v4.0.4. See [Release 4.0.4](https://github.com/webpro/knip/releases/tag/4.0.4) for release notes._

**Comment by webpro** [comment]
@uncenter Maybe hard to find  4️⃣0️⃣4️⃣  but you are still the center of v4.0.4 - thank you! 🙏

**Comment by uncenter** [comment]
No, thank you! Glad to help, I've been loving this tool. Working on another possible plugin at the moment to make Knip work even better for some Eleventy users!

**Comment by uncenter** [comment]
```
{ path: 'src/_includes/components/throbber.js', isDir: false }
neither, so added to copiedPackages
{ path: 'src/_includes/components/throbber.css', isDir: false }
neither, so added to copiedPackages
{ path: 'src/_includes/components/intl-number.js', isDir: false }
neither, so added to copiedPackages
{ path: 'node_modules/@11ty/logo/assets/logo-bg.svg', isDir: false }
neither, so added to copiedPackages
{ path: 'node_modules/@11ty/logo/assets/open-graph.jpg', isDir: false }
neither, so added to copiedPackages
{ path: 'node_modules/@11ty/logo/img/logo-784x1093.png', isDir: false }
neither, so added to copiedPackages
{ path: 'node_modules/@11ty/logo/img/logo-200x200.png', isDir: false }
neither, so added to copiedPackages
{ path: 'node_modules/@11ty/logo/img/logo-96x96.png', isDir: false }
neither, so added to copiedPackages
{
  path: 'node_modules/speedlify-score/speedlify-score.js',
  isDir: false
}
neither, so added to copiedPackages
{
  path: 'node_modules/@zachleat/seven-minute-tabs/seven-minute-tabs.js',
  isDir: false
}
neither, so added to copiedPackages
{
  path: 'node_modules/lite-youtube-embed/src/lite-yt-embed.js',
  isDir: false
}
neither, so added to copiedPackages
{
  path: 'node_modules/artificial-chart/artificial-chart.{css,js}',
  isDir: false
}
neither, so added to copiedPackages
{ path: 'netlify-email', isDir: true }
isDir so added plus glob to copiedEntries
{ path: 'src/img', isDir: true }
isDir so added plus glob to copiedEntries
{ path: 'src/blog/*.png', isDir: false }
neither, so added to copiedPackages
{ path: 'src/blog/pretty-atom-feed-v3.xsl', isDir: false }
neither, so added to copiedPackages
{ path: 'src/favicon.ico', isDir: false }
```

Hmm I feel like paths like `src/_includes/components/throbber.js` shouldn't be considered "packages". What is `isInternal` checking?

**Comment by uncenter** [comment]
This looks like it will fix it:

```js
if (isDir) {
    copiedEntries.add(join(path, `**/*.{${exts}}`));
} else if (isInNodeModules(path)) {
    copiedPackages.add(path);
} else {
    copiedEntries.add(path);
}
```

**Pull request merged**